lib/server.js

```diff
@@ -779,420 +700,250 @@ Server.prototype.toString = function toString() {
 
     str += sprintf(LINE_FMT, 'Accepts', this.acceptable.join(', '));
     str += sprintf(LINE_FMT, 'Name', this.name);
-    str += sprintf(LINE_FMT, 'Pre', handlersToString(this.before));
-    str += sprintf(LINE_FMT, 'Router', this.router.toString());
+    str += sprintf(
+        LINE_FMT,
+        'Pre',
+        handlersToString(this.preChain.getHandlers())
+    );
+    str += sprintf(LINE_FMT, 'Router', '');
+    this.router
+        .toString()
+        .split('\n')
+        .forEach(function forEach(line) {
+            str += sprintf('\t\t%s\n', line);
+        });
     str += sprintf(LINE_FMT, 'Routes', '');
-    Object.keys(this.routes).forEach(function forEach(k) {
-        var handlers = handlersToString(self.routes[k]);
-        str += sprintf(SUB_LINE_FMT, k, handlers);
+    _.forEach(this.router.getRoutes(), function forEach(route, routeName) {
+        var handlers = handlersToString(route.chain.getHandlers());
+        str += sprintf(SUB_LINE_FMT, routeName, handlers);
     });
     str += sprintf(LINE_FMT, 'Secure', this.secure);
     str += sprintf(LINE_FMT, 'Url', this.url);
-    str += sprintf(
-        LINE_FMT,
-        'Version',
-        Array.isArray(this.versions) ? this.versions.join() : this.versions
-    );
 
     return str;
 };
 
 ///--- Private methods
 
+// Lifecycle:
+//
+// 1. _onRequest (handle new request, setup request and triggers pre)
+//   2.1 _runPre
+//     2.2. _afterPre (runs after pre handlers are finisehd, triggers use)
+//       3.1 _runUse
+//         3.2. _afterUse (runs after use handlers are finised, triggers route)
+//          4.1 _runRoute
+//            4.2 _afterRoute (runs after last handler called next)
+//              4.3 _afterRoute (when, next('string') called)
+//              4.3 _routeErrorResponse (when error happaned in router)
+// 5. _finishReqResCycle (on response "finish" and "error" events)
```

Comment:
This waterfall is great! The additional granularity is nice. Is this order different than the current behavior? I thought the order is as follows:

1. pre
2. router.find
3. use (use only applies to routes that were successfully routed)
4. done/done(err)

Reply:
You are right, I was dumb enough to work based on my `express` and `koa` experience instead of reading the docs carefully :/

> Allows you to add in handlers that run for all routes. Note that handlers added via use() will run only after the router has found a matching route. If no match is found, these handlers will never run. - restify docs

*(In express/koa everything is a middleware, the router as well, so they don't make difference between pre/after route, it depends on the order of the middleware "registration")*

Btw maybe it's misleading for others as well coming from the `express` world, shouldn't we keep `use` as an alias and introduce a new term that reflects it's behavior better? Like `afterRoute()` or `useRoute()`, these are not very good names, but something like these :)